tests/test_qc_state_assignment.py

```diff
@@ -0,0 +1,234 @@
+from datetime import datetime
+
+import pytest
+from npg_id_generation.pac_bio import PacBioEntity
+from sqlalchemy import select
+
+from lang_qc.db.helper.qc import assign_qc_state_to_product, claim_qc_for_product
+from lang_qc.db.qc_schema import QcState, QcStateHist, SeqProduct, User
+from lang_qc.models.qc_state import QcStateBasic
+from lang_qc.util.errors import InconsistentInputError
+from tests.fixtures.well_data import (
+    load_data4qc_assign,
+    load_data4well_retrieval,
+    load_dicts_and_users,
+)
+
+time_now = datetime.now()
+
+
+def test_claim_qc(qcdb_test_session, load_data4well_retrieval, load_data4qc_assign):
+
+    users = qcdb_test_session.execute(select(User)).scalars().all()
+    user1 = users[0]
+    user2 = users[1]
+
+    id = PacBioEntity(
+        run_name="TRACTION_RUN_2", well_label="A1", plate_number=2
+    ).hash_product_id()
+    seq_product = qcdb_test_session.execute(
+        select(SeqProduct).where(SeqProduct.id_product == id)
+    ).scalar_one()
+
+    state_obj = claim_qc_for_product(
+        session=qcdb_test_session, seq_product=seq_product, user=user1
+    )
+    assert state_obj.date_created == state_obj.date_updated
+    assert state_obj.user == user1
+    assert state_obj.qc_type.qc_type == "sequencing"
+    assert state_obj.qc_state_dict.state == "Claimed"
+    assert state_obj.created_by == "LangQC"
+    assert state_obj.is_preliminary == 1
+    assert state_obj.seq_product.id_product == id
+    id_seq_product = state_obj.id_seq_product
+    hist_objs = _hist_objects(qcdb_test_session, id_seq_product)
+    assert len(hist_objs) == 1
+    _test_hist_object(state_obj, hist_objs[0])
+
+    # New QC state record is not created
+    state_obj = claim_qc_for_product(
+        session=qcdb_test_session, seq_product=seq_product, user=user2
+    )
+    assert state_obj.date_created == state_obj.date_updated
+    assert state_obj.user == user1
+    assert state_obj.qc_type.qc_type == "sequencing"
+    assert state_obj.qc_state_dict.state == "Claimed"
+    assert state_obj.created_by == "LangQC"
+    assert state_obj.is_preliminary == 1
+    assert state_obj.seq_product.id_product == id
+    assert state_obj.id_seq_product == id_seq_product
+    hist_objs = _hist_objects(qcdb_test_session, id_seq_product)
+    assert len(hist_objs) == 1
+
+
+def test_assign_qc_success(
+    qcdb_test_session, load_data4well_retrieval, load_data4qc_assign
+):
+
+    # The same well as claimed above.
+    id = PacBioEntity(
+        run_name="TRACTION_RUN_2", well_label="A1", plate_number=2
+    ).hash_product_id()
+
+    users = qcdb_test_session.execute(select(User)).scalars().all()
+    user2 = users[1]
+    seq_product = qcdb_test_session.execute(
+        select(SeqProduct).where(SeqProduct.id_product == id)
+    ).scalar_one()
+    id_seq_product = seq_product.id_seq_product
+    # In case oly this test is run, claim.
```

Comment:
only...

It is tricky to create a continuous test narrative with pytest since it allows for running individual tests functions. A whole test script might succeed, while test functions might fail if run individually if they rely on inheriting state from the previous test function.

The functions for claiming and assigning can be run with the same input repeatedly. If the qc_state object's attributes  are the same as the attributes of the current state of the same type, the current object is returned.

I'll try to rewrite this test so that it is self-sufficient.